test/unittests/eof_validation_test.cpp

```diff
@@ -58,145 +58,316 @@ TEST(eof_validation, validate_EOF_version)
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010001 030001 FE AA"),
+        EOFValidationError::unknown_section_id);
 }
 
 TEST(eof_validation, EOF1_truncated_section)
 {
-    EXPECT_EQ(validate_eof("EF0001 010002 00"), EOFValidationError::invalid_section_bodies_size);
-    EXPECT_EQ(validate_eof("EF0001 010002 00 FE"), EOFValidationError::invalid_section_bodies_size);
-    EXPECT_EQ(validate_eof("EF0001 010001 020002 00 FE"),
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010002 030000 00"),
+        EOFValidationError::invalid_section_bodies_size);
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010002 030000 00 000000"),
+        EOFValidationError::invalid_section_bodies_size);
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010002 030000 00 00000000 FE"),
+        EOFValidationError::invalid_section_bodies_size);
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010001 030002 00 00000000 FE"),
         EOFValidationError::invalid_section_bodies_size);
-    EXPECT_EQ(validate_eof("EF0001 010001 020002 00 FE AA"),
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010001 030002 00 00000000 FE AA"),
         EOFValidationError::invalid_section_bodies_size);
 }
 
+TEST(eof_validation, EOF1_code_section_offset)
+{
+    const auto eof =
+        "EF0001 010008 02000200030001 030004 00 00000001 00000000 6001fe fe 0000 0000"_hex;
+    ASSERT_EQ(validate_eof(EVMC_CANCUN, eof), EOFValidationError::success);
+
+    const auto header = read_valid_eof1_header(eof);
+    ASSERT_EQ(header.code_sizes.size(), 2);
+    EXPECT_EQ(header.code_sizes[0], 3);
+    EXPECT_EQ(header.code_sizes[1], 1);
+    ASSERT_EQ(header.code_offsets.size(), 2);
+    EXPECT_EQ(header.code_offsets[0], 25);
+    EXPECT_EQ(header.code_offsets[1], 28);
+}
+
 TEST(eof_validation, EOF1_trailing_bytes)
 {
-    EXPECT_EQ(validate_eof("EF0001 010001 00 FE DEADBEEF"),
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010001 030000 00 00000000 FE DEADBEEF"),
         EOFValidationError::invalid_section_bodies_size);
-    EXPECT_EQ(validate_eof("EF0001 010001 020002 00 FE AABB DEADBEEF"),
+    EXPECT_EQ(validate_eof("EF0001 010004 0200010001 030002 00 00000000 FE AABB DEADBEEF"),
         EOFValidationError::invalid_section_bodies_size);
 }
 
-TEST(eof_validation, EOF1_undefined_opcodes)
+TEST(eof_validation, EOF1_no_type_section)
 {
-    auto cont = "EF0001 010002 00 0000"_hex;
+    EXPECT_EQ(validate_eof("EF0001 0200010001 00 FE"),
+        EOFValidationError::code_section_before_type_section);
+    EXPECT_EQ(validate_eof("EF0001 02000200010001 00 FE FE"),
+        EOFValidationError::code_section_before_type_section);
+}
+
+TEST(eof_validation, EOF1_multiple_type_sections)
+{
+    EXPECT_EQ(validate_eof("EF0001 010004 010004 02000200010001 00 00000000 00000000 FE FE"),
+        EOFValidationError::multiple_type_sections);
+
+    //    Section order is must be (Types, Code+, Data)
+    //    EXPECT_EQ(validate_eof("EF0001 030002 010001 010001 030002 00 0000 FE FE 0000"),
```

Comment:
Why is this disabled?

Reply:
I will update the test to check different error. It fails earlier since we have different mandatory order of sections.
The error now is `data_section_before_types_section`